libraries/AP_BattMonitor/AP_BattMonitor_AD7091R5.cpp

```diff
@@ -0,0 +1,236 @@
+    AP_GROUPINFO("AMP_OFFSET", 34, AP_BattMonitor_AD7091R5, _curr_amp_offset, 0),
+
+    // @Param: VLT_OFFSET
+    // @DisplayName: Volage offset
+    // @Description: Voltage offset on voltage pin. This allows for an offset due to a diode. This voltage is subtracted before the scaling is applied
+    // @Units: V
+    // @User: Advanced
+    AP_GROUPINFO("VLT_OFFSET", 35, AP_BattMonitor_AD7091R5, _volt_offset, 0),
+    
+    // Param indexes must be less than 10 to avoid conflict with other battery monitor param tables loaded by pointer
+
+    AP_GROUPEND
+};
+
+AP_BattMonitor_AD7091R5::AnalogData AP_BattMonitor_AD7091R5::_analog_data[AD7091R5_NO_OF_CHANNELS] = {0};
+bool AP_BattMonitor_AD7091R5::_first = true;
+
+/**
+ * @brief Construct a new ap battmonitor ad7091r5::ap battmonitor ad7091r5 object
+ * 
+ * @param mon 
+ * @param mon_state 
+ * @param params 
+ */
+AP_BattMonitor_AD7091R5::AP_BattMonitor_AD7091R5(AP_BattMonitor &mon,
+                                             AP_BattMonitor::BattMonitor_State &mon_state,
+                                             AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+/**
+ * @brief probe and initialize the sensor and register call back
+ * 
+ */
+void AP_BattMonitor_AD7091R5::init(){
+    // voltage and current pins from params and check if there are in range
+    if (_volt_pin >= 1 && _volt_pin <= AD7091R5_NO_OF_CHANNELS) {
+        volt_buff_pt = _volt_pin.get() -1;
+    }
+    if (_curr_pin >= 1 && _curr_pin <= AD7091R5_NO_OF_CHANNELS) {
+        curr_buff_pt = _curr_pin.get() -1;
+    }
+
+    if (_first) { // only the first instance read the i2c device
+        _first = false;
+        _i2c_polling = true;
+        _state.healthy = false;
+        // probe i2c device
+        _dev = hal.i2c_mgr->get_device(AD7091R5_I2C_BUS, AD7091R5_I2C_ADDR);
+        if (_dev) {
+            WITH_SEMAPHORE(_dev->get_semaphore());
+            _dev->set_retries(10); // lots of retries during probe
+            //config
+            if (_reset() && _config()) {
+                _state.healthy = true;
+                _dev->set_retries(2); // drop to 2 retries for runtime
+                _dev->register_periodic_callback(AD7091R5_PERIOD_USEC, FUNCTOR_BIND_MEMBER(&AP_BattMonitor_AD7091R5::_timer, void));
+            }
+        }
+    }
+}
+
+/**
+ * config the adc
+ * - command mode
+ * - use external 3.3 reference
+ * - all channels enabled
+ * - set address pointer register to read the adc results
+ */
+bool AP_BattMonitor_AD7091R5::_config(){
+    uint8_t data[6] = {AD7091R5_CONF_ADDR, reg_conf_h, reg_conf_l, AD7091R5_CHAN_ADDR, reg_chan, AD7091R5_RESULT_ADDR};
+    return _dev->transfer(data, sizeof(data), nullptr, 0);
+}
+
+/**
+ * @brief read - read the voltage and curren
+ * 
+ */
+void AP_BattMonitor_AD7091R5::read(){
+    
+    WITH_SEMAPHORE(sem);
```

Comment:
We should be checking that any data was actually read off the bus within a reasonable time frame, and if not be setting the healthy flag to false, and maybe not update the used current calculation. It's quite plausible for an external I2C battery to just be lost for a bit due to EMI (particularly depending on how it's routed), and not using stale data is pretty important.